ms2deepscore/data_generators.py

```diff
@@ -171,46 +166,52 @@ def on_epoch_end(self):
+    @staticmethod
+    def _exclude_nans_from_labels(labels_df: pd.DataFrame):
+        """Exclude nans in labels_df, exclude columns and rows if there is any NaN
+        value"""
+        clean_df = labels_df.dropna(axis='rows')  # drop rows with any NaN
+        clean_df = clean_df[clean_df.index]  # drop corresponding columns
+        n_dropped = len(labels_df) - len(clean_df)
+        if n_dropped > 0:
+            print(f"{n_dropped} nans among {len(labels_df)} labels will be excluded.")
+        return clean_df
+
+    def _find_match_in_range(self, inchikey1, target_score_range, max_range=0.4):
         """Randomly pick ID for a pair with inchikey_id1 that has a score in
         target_score_range. When no such score exists, iteratively widen the range
         in steps of 0.1 until a max of max_range. If still no match is found take
         a random ID.
         Parameters
         ----------
-        inchikey_id1
-            ID of inchikey to be paired up with another compound within target_score_range.
+        inchikey1
+            Inchikey to be paired up with another compound within target_score_range.
         target_score_range
             lower and upper bound of label (score) to find an ID of.
         """
         # Part 1 - find match within range (or expand range iteratively)
         extend_range = 0
         low, high = target_score_range
+        inchikey2 = None
         while extend_range < max_range:
-            idx = np.where((self.score_array[inchikey_id1, self.inchikey_ids] > low - extend_range)
-                           & (self.score_array[inchikey_id1, self.inchikey_ids] <= high + extend_range))[0]
+            matching_inchikeys = self.labels_df.index[
+                (self.labels_df[inchikey1] > low - extend_range)
+                & (self.labels_df[inchikey1] <= high + extend_range)]
             if self.settings["ignore_equal_pairs"]:
-                idx = idx[idx != inchikey_id1]
-            if len(idx) > 0:
-                inchikey_id2 = self.inchikey_ids[np.random.choice(idx)]
+                matching_inchikeys = matching_inchikeys[matching_inchikeys != inchikey1]
+            if len(matching_inchikeys) > 0:
+                inchikey2 = np.random.choice(matching_inchikeys)
                 break
             extend_range += 0.1
 
-        # Part 2 - if still no match is found take the 2nd-highest score ID
-        if len(idx) == 0:
-            second_highest_id = self.score_array[inchikey_id1, np.array([x for x in self.inchikey_ids if x != inchikey_id1])].argmax()
-            if second_highest_id > inchikey_id1:
-                second_highest_id += 1
-            inchikey_id2 = self.inchikey_ids[second_highest_id]
+        # Part 2 - if still no match is found take the inchikey that has highest similarity score
+        # with inchikey1
+        # TODO: Why are we taking the highest score here? Shouldn't we take the one that is closest
```

Comment:
@florian-huber what do you think?

Reply:
My thinking behind taking the highest non-identical inchikey:
It sometimes happens that there is no Inchikey for a selected score-bin. Due to the strong bias in the data towards low similarity scores this only occurs for the higher bins (say >0.5). A perfect balancing of the data is hence not possible by the method provided here. The extending the range part ("Part 1") should compensate most of this. I didn't spend much time on this and simply felt like biasing the choice in "Part 2" towards highest-possible pairs could help as well...  but your solution seems a bit cleaner. So I am happy keeping your changes here (I expect that in practice this should often come down to the same choice).

Reply:
I've opened #32 for this.